expression/builtin_cast.go

```diff
@@ -292,7 +292,14 @@ func (c *castAsStringFunctionClass) getFunction(ctx sessionctx.Context, args []E
 	switch argTp {
 	case types.ETInt:
 		if bf.tp.GetFlen() == types.UnspecifiedLength {
-			bf.tp.SetFlen(args[0].GetType().GetFlen())
+			// check https://github.com/pingcap/tidb/issues/44786
+			// set flen from integers may truncate integers, e.g. char(1) can not display -1[int(1)]
+			// set it to 11 as mysql
+			if args[0].GetType().GetType() == mysql.TypeLong {
+				bf.tp.SetFlen(11)
+			} else {
+				bf.tp.SetFlen(args[0].GetType().GetFlen())
```

Comment:
Why not set it to `GetFlen()+1` ?

Reply:
`char(2)` can not display `-127 | int(1)`.